### ISSUE TITLE
Get specs updated to new dictionary defaulting setup

### ISSUE THREAD

**Issue by bzbarsky** [whatwg/webidl#758]
Followup to https://github.com/heycam/webidl/pull/750 that probably needs to wait for https://github.com/plinss/widlparser/issues/39 and https://github.com/tabatkins/bikeshed/issues/1491 to be fixed.

To-be-triaged list of things that might need updating, based on Gecko's IDL that needed to be changed:

- [ ] CSS -- https://github.com/w3c/csswg-drafts/issues/4234
  - [x] https://drafts.csswg.org/web-animations/
    - Animatable
    - AnimationEffect
    - AnimationPlaybackEvent
    - DocumentTimeline
    - KeyframeEffect
  - [x] AnimationEvent -- https://drafts.csswg.org/css-animations-1/
  - [x] https://drafts.csswg.org/cssom-view/
    - Element
    - GeometryUtils
    - MediaQueryListEvent
  - [x] https://drafts.csswg.org/css-font-loading/
    - FontFaceSetLoadEvent
    - FontFace
  - [ ] ResizeObserver -- https://drafts.csswg.org/resize-observer/
- [x] https://webaudio.github.io/web-audio-api/ -- https://github.com/WebAudio/web-audio-api/issues/1989
  - AnalyserNode
  - AudioBufferSourceNode
  - AudioContext
  - AudioWorkletNode
  - AudioWorkletProcessor
  - BaseAudioContext
  - BiquadFilterNode
  - ChannelSplitterNode
  - ChannelMergerNode
  - ConstantSourceNode
  - ConvolverNode
  - DelayNode
  - DynamicsCompressorNode
  - GainNode
  - MediaStreamAudioDestinationNode
  - OscillatorNode
  - PannerNode
  - PeriodicWave
  - StereoPannerNode
  - WaveShaperNode
- [x] https://w3c.github.io/ServiceWorker/ -- w3c/ServiceWorker#1448
  - CacheStorage
  - Cache
  - Clients
  - Client
  - ExtendableEvent
  - ExtendableMessageEvent
  - ServiceWorkerContainer
  - ServiceWorker
- [x] BlobEvent -- https://w3c.github.io/mediacapture-record/
- [x] https://w3c.github.io/FileAPI/ -- https://github.com/w3c/FileAPI/pull/134
  - File
  - Blob
- [x] https://html.spec.whatwg.org/ - https://github.com/whatwg/html/pull/4745
  - CanvasRenderingContext2D
  - CloseEvent
  - CustomElementRegistry
  - DedicatedWorkerGlobalScope
  - DragEvent
  - Element
  - ErrorEvent
  - EventSource
  - HashChangeEvent
  - HTMLSlotElement
  - MessageEvent
  - MessagePort
  - PageTransitionEvent
  - PopStateEvent
  - SharedWorker
  - StorageEvent
  - TrackEvent
  - WindowOrWorkerGlobalScope
  - Window
  - Worker
- [x] ClipboardEvent -- https://w3c.github.io/clipboard-apis/ -- w3c/clipboard-apis#91
- [x] https://w3c.github.io/uievents/ -- w3c/uievents#239
  - CompositionEvent
  - FocusEvent
  - InputEvent
  - KeyboardEvent
  - MouseEvent
  - UIEvent
  - WheelEvent
- [x] CredentialManagement -- https://github.com/w3c/webappsec-credential-management/issues/140
- [x] https://dom.spec.whatwg.org/ -- https://github.com/whatwg/dom/pull/771
  - CustomEvent
  - Document
  - EventTarget
  - Event
  - MutationObserver
  - Node
- [x] https://w3c.github.io/deviceorientation/ -- https://github.com/w3c/deviceorientation/pull/78
  - DeviceMotionEvent
  - DeviceOrientationEvent
- [x] DeviceLightEvent -- https://github.com/w3c/ambient-light/pull/58
- [x] https://drafts.fxtf.org/geometry/ -- https://github.com/w3c/fxtf-drafts/issues/356
  - DOMMatrix
  - DOMPoint
  - DOMQuad
  - DOMRect
- [x] FileSystemDirectoryEntry -- https://wicg.github.io/entries-api/ -- https://github.com/WICG/entries-api/issues/31
- [x] GamepadEvent -- https://github.com/w3c/gamepad/issues/101
- [x] Geolocation -- https://www.w3.org/TR/geolocation-API/ -- https://github.com/w3c/geolocation-api/pull/27
- [x] https://w3c.github.io/IndexedDB/ -- https://github.com/w3c/IndexedDB/issues/285
  - IDBDatabase
  - IDBObjectStore
  - IDBVersionChangeEvent
- [x] ImageCaptureErrorEvent -- https://github.com/w3c/mediacapture-image/issues/214
- [x] IntersectionObserver -- w3c/IntersectionObserver#374
- [x] https://github.com/w3c/mediacapture-main/issues/613
  - MediaDevices
  - MediaStreamTrack
- [x] MediaEncryptedEvent -- https://github.com/w3c/encrypted-media/issues/452
- [x] MediaRecorder -- https://github.com/w3c/mediacapture-record/issues/180
- [ ] MediaStreamEvent -- no longer specified?
- [x] WebAudio/web-midi-api#203 
  - MIDIConnectionEvent
  - MIDIMessageEvent
  - Navigator.requestMIDIAccess
- [x] Navigator.share -- https://github.com/WICG/web-share/pull/111
- [x] https://notifications.spec.whatwg.org/ - https://github.com/whatwg/notifications/pull/149
  - [x] Notification
  - [x] ServiceWorkerRegistration
- [x] Payment Request -- w3c/payment-request#874
  - MerchantValidationEvent
  - PaymentMethodChangeEvent
  - PaymentRequestUpdateEvent
  - PaymentRequest
  - PaymentResponse
- [x] Performance -- https://github.com/w3c/user-timing/pull/65
- [x] PerformanceObserverEntryList -- https://github.com/w3c/performance-timeline/pull/61
- [x] PerformanceObserver -- https://github.com/w3c/performance-timeline/pull/147
- [x] PointerEvent -- https://github.com/w3c/pointerevents/issues/296
- [x] ProximitySensor -- https://github.com/w3c/proximity/issues/40
- [x] Push w3c/push-api#308
  - PushEvent
  - PushManager
  - PushSubscription
- [x] Reporting -- https://github.com/w3c/reporting/issues/172
- [x] https://fetch.spec.whatwg.org/ - https://github.com/whatwg/fetch/pull/919
  - [x] Request
  - [x] Response
- [x] https://github.com/w3c/webrtc-pc/issues/2277
  - RTCDTMFToneChangeEvent
  - RTCIceCandidate
  - RTCPeerConnectionIceEvent
  - RTCPeerConnection
  - RTCSessionDescription
- [ ] SecurityPolicyViolationEvent -- https://github.com/w3c/webappsec-csp/issues/408
- [x] https://github.com/w3c/speech-api/issues/60
  - SpeechRecognitionErrorEvent
  - SpeechRecognitionEvent
- [x] https://github.com/w3c/svgwg/issues/727
  - SVGGeometryElement
  - SVGGraphicsElement
  - SVGTextContentElement
- [x] TextDecoder -- https://encoding.spec.whatwg.org/ - https://github.com/whatwg/encoding/pull/181
- [x] TouchEvent -- https://github.com/w3c/touch-events/issues/104
- [x] TransitionEvent -- https://github.com/w3c/csswg-drafts/pull/4232
- [ ] WebGLContextEvent
- [x] WebGPU -- https://github.com/gpuweb/gpuweb/issues/420
- [x] Worklet -- https://github.com/w3c/css-houdini-drafts/issues/937
- [x] https://xhr.spec.whatwg.org/ - https://github.com/whatwg/xhr/pull/250
  - [x] ProgressEvent

**Comment by bzbarsky** [comment]
@annevk @domenic

**Comment by marcoscaceres** [comment]
I can help with the WebApps WG specs and Payments WG spec and get them republished. 

ReSpec will emit warnings for this too. We might also implement an “auto fix” for this in ReSpec, so at least Editors drafts will produce the right IDL.

**Comment by bzbarsky** [comment]
@marcoscaceres Thanks!  I'll keep triaging the list above as I have time, and trying to get issues filed.

**Comment by inexorabletash** [comment]
Once bikeshed is ready, I can take the IndexedDB and entries-api ones.

In the above checklist, MediaStreamAudioDestinationNode, OscillatorNode, PannerNode, StereoPannerNode, WaveShaperNode, PeriodicWave should be added to the webaudio batch.

Presumably the UDP* and TCP* ones are Firefox OS residue?

**Comment by bzbarsky** [comment]
> should be added to the webaudio batch

Done, thanks.  Hadn't gotten to those yet.  ;)

> Presumably the UDP* and TCP* ones are Firefox OS residue?

They started out with Firefox OS, but at this point it looks like Firefox devtools use them (e.g. for adb connection).  More pertinently, https://www.w3.org/2012/sysapps/tcp-udp-sockets/ is a thing, but doesn't seem to contain any actual IDL, though https://www.w3.org/TR/tcp-udp-sockets/ does.  Not sure whether it's worth trying to get that updated; it seems pretty dead at the moment...  I'll take all the TCP* and UDP* bits out of the list for now.

> Also, IDBFileHandle is non-standard, it's a Gecko thing

Yep, I just dropped all the IDB things in without checking carefully, since the IDB spec needs _some_ updates...  I'll remove it from the list.

> Similarly, there is no standard IDBFactory method that takes a dictionary - that's a nonstandard Gecko extension.

Removed from list, thanks.

There are likely more Gecko-specific things in the untriaged list; I removed some obvious ones as a first pass, but as you found there's more there...

Thank you for filing the IDB and entries-api issues!

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/payment-request/pull/874  for Payment  Request:

You can ✓:

-   PaymentMethodChangeEvent
-   PaymentRequestUpdateEvent
-   PaymentRequest
-   PaymentResponse

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/FileAPI/pull/134 for File API, which covers:
 
 * File
 * Blob

**Comment by CYBAI** [comment]
Sent w3c/ServiceWorker#1448 for ServiceWorker, which covers:

- ServiceWorker
- ServiceWorkerContainer
- Client
- Clients
- ExtendableEvent
- ExtendableMessageEvent
- Cache
- CacheStorage

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/push-api/pull/308 for Push, which covers: 

-   PushEvent
-   PushManager
-   PushSubscription

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/clipboard-apis/pull/91 for Clipboard APIs, which covers:

 * ClipboardEvent

**Comment by marcoscaceres** [comment]
Cc @anssiko... there are a few above from your WGs. Could use your help 🙏

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/IntersectionObserver/pull/374 for IntersectionObserver, which covers:

* IntersectionObserver

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/uievents/pull/239 for UIEvents, which covers:

 * UIEvent
 * FocusEvent
 * MouseEvent
 * WheelEvent
 * InputEvent
 * KeyboardEvent
 * CompositionEvent

**Comment by marcoscaceres** [comment]
Note above... `InputEvent` is defined in UIEvents, not in https://w3c.github.io/input-events/, where it's a partial interface.

**Comment by marcoscaceres** [comment]
Sent https://github.com/WebAudio/web-midi-api/pull/203 for Web Midi, which covers:

 - MIDIConnectionEvent
 - MIDIMessageEvent

**Comment by marcoscaceres** [comment]
@yoavweiss a few here from the Perf WG:

*  PerformanceEntryEvent
*  PerformanceObserverEntryList
*  PerformanceObserver

If you let me  know the specs, I can send some PRs.

**Comment by annevk** [comment]
All WHATWG standards have a PR at this point.

**Comment by anssiko** [comment]
PRs submitted for https://github.com/w3c/deviceorientation/pull/78 and  https://github.com/w3c/geolocation-api/pull/27. DeviceProximityEvent, UserProximityEvent, and DeviceLightEvent are deprecated and superseded by ProximitySensor and AmbientLightSensor that were affected and received a similar PR treatment.

In addition, I created PRs for specs that extend Generic Sensor API, also affected.

**Comment by Ms2ger** [comment]
Everything except WebGL should be filed now; I also removed the Mozilla-internal APIs.

**Comment by domenic** [comment]
Awesome! How did you feel about this process? In particular, how painful was it, and how does that make us feel about changes like #700, or other more speculative ideas people have thrown around?

**Comment by annevk** [comment]
My main takeaway is that it'd be great to have all tooling PRs ready as well so we only need to communicate a change to everyone once. So that at the point it's communicated everyone can pick it up straight away.

**Comment by Ms2ger** [comment]
It wasn't too bad, I think. @bzbarsky's list was very helpful.

**Comment by anssiko** [comment]
Second that, the list was helpful.

Related to tooling, and hopefully not too much OT, do we have a tool to conveniently diff spec IDLs against major browsers' implementations to make communicating (proposed) spec changes not just toward downstream specs but also toward implementations easier?

There's https://github.com/mdittmer/web-apis and https://github.com/tidoust/reffy so parts of the problem have already been solved. @foolip @tidoust?

**Comment by foolip** [comment]
@anssiko a part of https://github.com/mdittmer/web-apis was indeed directed towards this problem, and @mdittmer did in fact produce lists of discrepancies in various categories. We used these lists to file a lot of bugs tracked by https://crbug.com/674593. You can see a lot have been resolved, so the approach worked!

However, we didn't keep triaging differences over time or build tooling to prevent differences from occurring in the first place.

The biggest hurdle, which isn't a very big one, is having a Web IDL parser able to parse both all spec IDL (webidl2.js) but also all the IDL from Chromium, Gecko and WebKit. https://github.com/mdittmer/webidl2-js (not the same as webidl2.js) was once able to do that, but such a parser has to be continuously maintained to keep up with changes.

So to summarize, such tooling isn't deployed, but it's been done before, and the technical challenges are pretty modest. It's mostly a matter of time investment vs. payoff, and web-platform-tests took priority for us.

@lukebjerring FYI about this history and potential area of work.

**Comment by annevk** [comment]
I'm closing this. There's one or two specs still not updated, but by-and-large this has been embraced and tooling will eventually get the laggards updated as well.
